### DISCUSSION TITLE
How to build a package and its depency from a package.json script, cross platform ?

### DISCUSSION THREAD

**Discussion by stevebeauge** [vercel/turborepo#1557] [Help] [answered]
I'm working in a large monorepo that basically consists in 3 packages categories : tooling, common code and apps.

In the "apps" package, I want to add a "build" script in the package.json that will build the app and its dependencies.

Because I have a more complex build chain, I want to split building the dependencies from the app itself.

In summary I have:
* (root)
  * package.json (which contains script to call turbo, because turbo must be run from root) 
  * common/shared
    * package.json (which contains a simple build script that will call tsc)
  * apps/myapp
     * package.json (contains build scripts and dependency to common/shared

In "myapp"'s package.json I defined:

````json
        "build:dependencies": "pnpm -w run turbo --filter=\"@monorepo/myapp^...\"",
        "build": "pnpm run build:dependencies && gulp bundle",
````

As you can see, I expect to build the dependencies from a dedicated script then the app itself.

This approach works well (while the syntax is actually tricky to define), but not on Windows. Specifically, the `^` must be doubled on windows computers (this is known issue: https://github.com/npm/npm/issues/17692, never fixed).

To make it works on Windows, I have to fix the build like this:

````json
        "build:dependencies": "pnpm -w run turbo --filter=\"@monorepo/myapp^^^^...\"",
        "build": "pnpm run build:dependencies && gulp bundle",
````

I have actually 4 carets, because I "send" the filter to the script at the workspace root. SO I have to double the caret for each indirection.

This works fine on windows, but not on *nix systems (WSL especially).

The only way I found is to declare something like: 

````json
        "build:dependencies": "pnpm -w run turbo --filter=\"@monorepo/myapp^^^^...\"",
        "build": "pnpm run build:dependencies && gulp bundle",
        "build:dependencies:wsl": "pnpm -w run turbo --filter=\"@monorepo/myapp^...\"",
        "build:wsl": "pnpm run build:dependencies:wsl && gulp bundle",
````

But it's not very satisfying.

Is there anything I can do to fix this ?

I know the root cause is not turbo's fault, but I hope someone has a workaround.

PS: if it matters, my root package.json define this script :

````json
    "scripts": {
        "turbo": "turbo run build"
    },
````

Thanks to pnpm which pass arguments to scripts, I can call turbo from anywhere, providing what I want to build.

**Comment by anthonyshew** [comment, marked as answer]
Your setup sounds reasonable to me, given your build-chain complexity.

Nix and WSL are not on our listed [Support Policy](https://turbo.build/repo/docs/support#support-policy).
